Common/include/toolboxes/CInletInterpolation.hpp

```diff
@@ -0,0 +1,84 @@
+/*!
+ * \file inlet_interpolation_functions.hpp
+ * \brief Inlet_interpolation_functions
+ * \author Aman Baig
+ * \version 7.0.0 "Blackbird"
+ *
+ * SU2 Project Website: https://su2code.github.io
+ *
+ * The SU2 Project is maintained by the SU2 Foundation 
+ * (http://su2foundation.org)
+ *
+ * Copyright 2012-2019, SU2 Contributors (cf. AUTHORS.md)
+ *
+ * SU2 is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * SU2 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with SU2. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <iostream>
+#include <cmath>
+#include <vector>
+#include<fstream>
+#include "../datatype_structure.hpp"
+
+using namespace std;
+
+class CAkimaSpline{ 
+protected:
+vector<su2double> x,y,b,c,d;
+int n;
+public:
+void akima_spline_set(CAkimaSpline *s, vector<su2double> &x, vector<su2double> &y);
+su2double akima_spline_eval(CAkimaSpline *s, su2double Point_Interp);
+};
+
+class CLinearSpline{
+    protected:
+    vector<su2double> x,y,dydx;
+    public:
+    void Linear_set(CLinearSpline *s,vector<su2double> &x, vector<su2double> &y);
+    su2double Linear_eval(CLinearSpline *s, su2double Point_Interp);
+    bool Point_Match = false;
+};
+
+class CInletInterpolation:public CAkimaSpline,public CLinearSpline{
+string Interpolation_Type, Interpolation_Function;
+public:
+void Interpolation_Set(CInletInterpolation *s, vector<su2double> &x, vector<su2double> &y, string Interpolation_Function){
+    if(Interpolation_Type == "Akima")
+        return s->akima_spline_set(s,x,y);
+    else if(Interpolation_Type == "Linear")
+        return s->Linear_set(s,x,y);
+}
```

Comment:
If I turn the interpolation inheritance upside down then I would need to make separate objects for the interpolation class I want to do. For Akima I would need to call 'CAkimaSpline' and for Linear 'CLinearSpline'. Right now I call only CInletInterpolation and pass a string (or enum) to let it know what to do. Don't you think the latter way is simpler?

I could include a constructor in CInletInterpolation to let it know which class to initialize from using the enum option passed.

I do need to pass the pointer to the Interpolator to let it know which column to interpolate (which interpolation function to use). I have included the details for this in my other comments.
https://github.com/su2code/SU2/blob/c29a8f042c3af802e8bcb8e3228bb2b900fad46c/SU2_CFD/src/solvers/CSolver.cpp#L4314

I will rename CInletInterpolation to C1DInterpolation.

Reply:
No it is not simpler, that line you show there `s[iCol]->Interpolation_Set(s[iCol],...` looks like trouble from a mile away, you have a method of an object getting the object itself as an argument. There situations where something like this might be needed, but this is not one of them, if you turn the inheritance the right way around this is what your code should look like.
```c++
C1DInterpolation* interpolator = nullptr; // use descriptive names for your variables
switch(InterpolationType) {
  case LINEAR_INTEPOLATION:
    interpolator = new CLinearInterpolation(InletData,nRows,nCols,...);
    break;
  case AKIMA_INTERPOLATION:
    interpolator = new CAkimaInterpolation(InletData,nRows,nCols,...);
    break;
  default:
    SU2_MPI::Error("...
    break;
}

// Do stuff with the interpolator
interpolator->

// Delete the interpolator
delete interpolator;
```

Reply:
This does look very clean. However, I would have to separate the columns in each of the constructors. The "set" and "eval" functions can only handle one set of piece-wise interpolation functions (only one column). In some way I need to make more instances of them to allow for more columns. 
If you don't like this way of calling:
`Inlet_Interpolated[iVar]=s[iVar]->Interpolation_Evaluate(s[iVar],Interp_Radius,Interpolation_Function);`
can I use the same approach but use this-> pointer to access the object data rather than passing it as an argument?

In either case I would need to do something like this (matrix of interpolators):
`             CInletInterpolation **s = new CInletInterpolation*[nRows];
              for (int i=0;i<=nColumns;i++)
                s[i] = new CInletInterpolation[nRows];`